legate/driver/main.py

```diff
@@ -60,3 +62,46 @@ def main(argv: list[str]) -> int:
         raise e
 
     return driver.run()
+
+
+def canonical_main(argv: list[str]) -> Tuple[Tuple[str, ...], Dict[str, str]]:
+    """A main function for the canonical driver that can be used
+    programmatically or by entry-points.
+
+    Parameters
+    ----------
+        argv : list[str]
+            Command-line arguments to start the canonical driver with
+
+    Returns
+    -------
+        Tuple[Tuple[str], Dict[str, str]]
+            Legion command and env of the driver
+
+    """
+    from ..util.system import System
+    from ..util.ui import error
+    from . import CanonicalDriver, Config
+    from .driver import print_verbose
+
+    try:
+        config = Config(argv)
+    except Exception as e:
+        print(error("Could not configure canonical driver:\n"))
+        raise e
+
+    try:
+        system = System()
+    except Exception as e:
+        print(error("Could not determine System settings: \n"))
+        raise e
+
+    try:
+        driver = CanonicalDriver(config, system)
+    except Exception as e:
+        msg = "Could not initialize canonical driver, path config and exception follow:"  # noqa
+        print(error(msg))
+        print_verbose(system)
+        raise e
+
+    return driver.cmd, driver.env
```

Comment:
I'm not sure this function belongs here and I am not sure it should be called a "main" since main function are typically responsible for ultimately running something as an "entry point", and this function does not do that. Maybe `prepare_legion_main_args` or similar would be more descriptive?

Reply:
A possible factorization:

```python
def prepare_driver(driver_type):
    from ..util.system import System
    from ..util.ui import error
    from . import Config
    from .driver import print_verbose

    try:
        config = Config(argv)
    except Exception as e:
        print(error("Could not configure canonical driver:\n"))
        raise e

    try:
        system = System()
    except Exception as e:
        print(error("Could not determine System settings: \n"))
        raise e

    try:
        driver = driver_type(config, system)
    except Exception as e:
        msg = "Could not initialize driver, path config and exception follow:"  # noqa
        print(error(msg))
        print_verbose(system)
        raise e

    return driver
````

This function could be called by `main` (passing in a plain `Driver`) and also used up in `__init__` (passing in a `CanonicalDriver`)

This is also another case we might want to rename `Driver` to something more specific like `LegateDriver` now that we are adding another specialized kind of driver alongside.

Reply:
I have renamed the `Driver` to `LegateDriver`, `main` to `legate_main`.  I also unified `legate_main` with `canonical_main` by using the `prepare_driver`, so now both `legate_main` and `canonical_main` calls `prepare_driver` to create a `LegateDriver` or `CanonicalDriver`.